src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -525,6 +525,135 @@ public override BoundNode VisitTypeOrInstanceInitializers(BoundTypeOrInstanceIni
             return new BoundStatementList(node.Syntax, rewrittenStatements, node.HasErrors);
         }
 
+        public override BoundNode VisitArrayAccess(BoundArrayAccess node)
+        {
+            // If the array access index is of type System.Index or System.Range
+            // we need to translate it into a call into an actual indexer, instead
+            // of a simple array element access.
+
+            if (this.EmitModule is null || 
+                node.Indices.Length != 1 ||
+                node.Indices[0].Type.SpecialType != SpecialType.None)
+            {
+                return base.VisitArrayAccess(node);
+            }
+
+            // PROTOTYPE: THIS MUST BE REMOVED BEFORE DEV16
+            // Since it takes a long time to update mscorlib we fake up the right
+            // code to pretend there are two indexers that support Index and Range
+            // on arrays.
+
+            var syntax = node.Syntax;
+            var F = _factory;
+            var indexLocal = F.StoreToTemp(
+                VisitExpression(node.Indices[0]),
+                out BoundAssignmentOperator indexAssign);
+            var arrayLocal = F.StoreToTemp(
+                VisitExpression(node.Expression),
+                out BoundAssignmentOperator arrayAssign);
+            var indexType = VisitType(node.Indices[0].Type);
+
+            var indexValueSymbol = (PropertySymbol)F.WellKnownMember(WellKnownMember.System_Index__Value);
+            var indexFromEndSymbol = (PropertySymbol)F.WellKnownMember(WellKnownMember.System_Index__FromEnd);
```

Comment:
I don't believe that we verify these are available anywhere in the change and hence could be `null`. Shouldn't we verify these exist at the point we allow the indexer to be bound in the Binder?

Reply:
factory.WellKnownMember checks for the presence of the member and adds a diagnostic if it's not present.

Reply:
(This is done by throwing an exception that's caught by the local rewriter).